# Classification Exercises

In [1]:
from env import host, password, user
import pandas as pd
import numpy as np
import os
import seaborn as sns
import matplotlib.pyplot as plt
from env import host, user, password
from pydataset import data

The end product of this exercise is a jupyter notebook (classification_exercises.ipynb) and acquire.py. The notebook will contain all your work as you move through the exercises. The acquire.py file should contain the final functions.

1. In a jupyter notebook, classification_exercises.ipynb, use a python module (pydata or seaborn datasets) containing datasets as a source from the iris data. Create a pandas dataframe, df_iris, from this data.
 - print the first 3 rows
 - print the number of rows and columns (shape)
 - print the column names
 - print the data type of each column
 - print the summary statistics for each of the numeric variables. Would you recommend rescaling the data based on these statistics?

In [2]:
df_iris = data('iris')

In [3]:
# print the first 3 rows
df_iris.head(3)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
1           5.1          3.5           1.4          0.2  setosa
2           4.9          3.0           1.4          0.2  setosa
3           4.7          3.2           1.3          0.2  setosa

In [4]:
# print the number of rows and columns (shape)
df_iris.shape

(150, 5)

In [5]:
# print the column names
df_iris.columns

Index(['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width',
       'Species'],
      dtype='object')

In [6]:
# print the data type of each column
df_iris.dtypes

Sepal.Length    float64
Sepal.Width     float64
Petal.Length    float64
Petal.Width     float64
Species          object
dtype: object

In [7]:
# print the summary statistics for each of the numeric variables. Would you recommend rescaling the data based on these statistics?
df_iris.describe()
# No I wouldn't recommend rescaling.

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [28]:
df_iris.describe().T

count      mean       std  min  25%   50%  75%  max
Sepal.Length  150.0  5.843333  0.828066  4.3  5.1  5.80  6.4  7.9
Sepal.Width   150.0  3.057333  0.435866  2.0  2.8  3.00  3.3  4.4
Petal.Length  150.0  3.758000  1.765298  1.0  1.6  4.35  5.1  6.9
Petal.Width   150.0  1.199333  0.762238  0.1  0.3  1.30  1.8  2.5

2. Read the Table1_CustDetails table from the Excel_Exercises.xlsx file into a dataframe named df_excel.
 - assign the first 100 rows to a new dataframe, df_excel_sample
 - print the number of rows of your original dataframe
 - print the first 5 column names
 - print the column names that have a data type of object
 - compute the range for each of the numeric variables.

In [8]:
df_excel = pd.read_excel('Excel_Exercises.xlsx', sheet_name='Table1_CustDetails')

In [9]:
# assign the first 100 rows to a new dataframe, df_excel_sample
df_excel_sample = df_excel.head(100)

In [10]:
# print the number of rows of your original dataframe
df_excel.shape

(7049, 12)

In [11]:
df_excel.shape[0]

7049

In [12]:
# print the first 5 column names
df_excel.columns[:5]

Index(['customer_id', 'gender', 'is_senior_citizen', 'partner', 'dependents'], dtype='object')

In [13]:
# print the column names that have a data type of object
df_excel.select_dtypes('object')
df_excel.select_dtypes('object').columns

Index(['customer_id', 'gender', 'partner', 'dependents', 'payment_type',
       'churn'],
      dtype='object')

In [14]:
# compute the range for each of the numeric variables.
df_excel.dtypes

customer_id           object
gender                object
is_senior_citizen      int64
partner               object
dependents            object
phone_service          int64
internet_service       int64
contract_type          int64
payment_type          object
monthly_charges      float64
total_charges        float64
churn                 object
dtype: object

In [15]:
df_excel.select_dtypes(['int64', 'float64']).max() - df_excel.select_dtypes(['int64', 'float64']).min()

is_senior_citizen       1.0
phone_service           2.0
internet_service        2.0
contract_type           2.0
monthly_charges       100.5
total_charges        8666.0
dtype: float64

In [16]:
df_excel.select_dtypes(exclude='object').max() - df_excel.select_dtypes(exclude='object').min()

is_senior_citizen       1.0
phone_service           2.0
internet_service        2.0
contract_type           2.0
monthly_charges       100.5
total_charges        8666.0
dtype: float64

3. Read the data from this google sheet into a dataframe, df_google
 - print the first 3 rows
 - print the number of rows and columns
 - print the column names
 - print the data type of each column
 - print the summary statistics for each of the numeric variables
 - print the unique values for each of your categorical variables

In [17]:
# df_google = csv_export_url = sheet_url.replace('/edit?usp=sharing', '/export?format=csv&gid=341089357')

In [18]:
# print the first 3 rows
sheet_url = 'https://docs.google.com/spreadsheets/d/1Uhtml8KY19LILuZsrDtlsHHDC9wuDGUSe8LTEwvdI5g/edit#gid=341089357'    

csv_export_url = sheet_url.replace('/edit#gid=', '/export?format=csv&gid=')

df_google = pd.read_csv(csv_export_url)
df_google.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Thayer)  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [19]:
# print the number of rows and columns
df_google.shape

(891, 12)

In [20]:
# print the column names
df_google.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [21]:
# print the data type of each column
df_google.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [22]:
# print the summary statistics for each of the numeric variables
df_google.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [23]:
# print the unique values for each of your categorical variables
df_google.select_dtypes(include='object').nunique()

Name        891
Sex           2
Ticket      681
Cabin       147
Embarked      3
dtype: int64

Make a new python module, acquire.py to hold the following data aquisition functions:

1. Make a function named get_titanic_data that returns the titanic data from the codeup data science database as a pandas data frame. Obtain your data from the Codeup Data Science Database.

In [24]:
from acquire import get_titanic_data

In [25]:
df = get_titanic_data()
df.head(2)

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    male  22.0      1      0   7.2500   
1             1         1       1  female  38.0      1      0  71.2833   

  embarked  class  deck  embark_town  alone  
0        S  Third  None  Southampton      0  
1        C  First     C    Cherbourg      0

In [26]:
df = get_titanic_data(cached=True)
df.head(2)

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    male  22.0      1      0   7.2500   
1             1         1       1  female  38.0      1      0  71.2833   

  embarked  class deck  embark_town  alone  
0        S  Third  NaN  Southampton      0  
1        C  First    C    Cherbourg      0

2. Make a function named get_iris_data that returns the data from the iris_db on the codeup data science database as a pandas data frame. The returned data frame should include the actual name of the species in addition to the species_ids. Obtain your data from the Codeup Data Science Database.

In [65]:
from acquire import get_iris_data

In [66]:
iris_df = get_iris_data()
iris_df.head()

species_id  measurement_id  sepal_length  sepal_width  petal_length  \
0           1               1           5.1          3.5           1.4   
1           1               2           4.9          3.0           1.4   
2           1               3           4.7          3.2           1.3   
3           1               4           4.6          3.1           1.5   
4           1               5           5.0          3.6           1.4   

   petal_width species_name  
0          0.2       setosa  
1          0.2       setosa  
2          0.2       setosa  
3          0.2       setosa  
4          0.2       setosa

3. Once you've got your get_titanic_data and get_iris_data functions written, now it's time to add caching to them. To do this, edit the beginning of the function to check for a local filename like titanic.csv or iris.csv. If they exist, use the .csv file. If the file doesn't exist, then produce the SQL and pandas necessary to create a dataframe, then write the dataframe to a .csv file with the appropriate name.

# Data Preparation Exercises

The end product of this exercise should be the specified functions in a python script named prepare.py. Do these in your classification_exercises.ipynb first, then transfer to the prepare.py file.

This work should all be saved in your local classification-exercises repo. Then add, commit, and push your changes.

Using the Iris Data:

1. Use the function defined in acquire.py to load the iris data.

In [54]:
from acquire import get_iris_data

In [71]:
iris_df.head()

species_id  measurement_id  sepal_length  sepal_width  petal_length  \
0           1               1           5.1          3.5           1.4   
1           1               2           4.9          3.0           1.4   
2           1               3           4.7          3.2           1.3   
3           1               4           4.6          3.1           1.5   
4           1               5           5.0          3.6           1.4   

   petal_width species_name  
0          0.2       setosa  
1          0.2       setosa  
2          0.2       setosa  
3          0.2       setosa  
4          0.2       setosa

2. Drop the species_id and measurement_id columns.

In [56]:
columns_to_drop = ['species_id', 'measurement_id']

In [57]:
iris_df = iris_df.drop(columns=columns_to_drop)

In [58]:
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species_name
0           5.1          3.5           1.4          0.2       setosa
1           4.9          3.0           1.4          0.2       setosa
2           4.7          3.2           1.3          0.2       setosa
3           4.6          3.1           1.5          0.2       setosa
4           5.0          3.6           1.4          0.2       setosa

3. Rename the species_name column to just species.

In [59]:
iris_df = iris_df.rename(columns= {'species_name':'species'})
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

4. Create dummy variables of the species name.

In [61]:
dummy_iris_df = pd.get_dummies(iris_df[['species',]], dummy_na=False, drop_first=[True])
dummy_iris_df.head()

species_versicolor  species_virginica
0                   0                  0
1                   0                  0
2                   0                  0
3                   0                  0
4                   0                  0

5. Create a function named prep_iris that accepts the untransformed iris data, and returns the data with the transformations above applied.

In [76]:
def prep_iris(df):
    '''
    takes in a dataframe of the iris dataset as it is acquired and returns a cleaned dataframe
    arguments: df: a pandas DataFrame with the expected feature names and columns
    return: clean_df: a dataframe with the cleaning operations performed on it
    '''
    df = df.drop_duplicates()
    df = df.drop(columns=['species_id', 'measurement_id'])
    df = df.rename(columns= {'species_name':'species'})
    dummy_iris_df = pd.get_dummies(df[['species',]], dummy_na=False, drop_first=[True])
    df = pd.concat([df, dummy_iris_df], axis=1)
    return df.drop(columns=['species'])

In [77]:
cleaned_df_v0 = prep_iris(iris_df)

In [79]:
cleaned_df_v0.head()

sepal_length  sepal_width  petal_length  petal_width  species_versicolor  \
0           5.1          3.5           1.4          0.2                   0   
1           4.9          3.0           1.4          0.2                   0   
2           4.7          3.2           1.3          0.2                   0   
3           4.6          3.1           1.5          0.2                   0   
4           5.0          3.6           1.4          0.2                   0   

   species_virginica  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

## Bonus Exercises
2. Prepare Titanic Data - Challenge/Bonus Exercises

- 2.1  Use the function you defined in acquire.py to load the titanic dataset.
- 2.2  Handle the missing values in the embark_town and embarked columns.
- 2.3  Remove the deck column.
- 2.4  Create a dummy variable of the embarked column.
- 2.5  Split Data
- 2.5.1  Create a helper function to split titanic data into train, validate, test datasets
- 2.6  Fill or impute the missing values in age.
- 2.6.1  Build a helper function for imputing
- 2.7  Create a prep_titanic function that accepts the untransformed titanic data and returns the data with the transformations above applied.